SDK 2.0

WoLfulus opened the pull request:
# Overview

(reopened this PR from a different branch)

- Makes folder independent from the root package.json/lerna and can be developed on its own
- Separate `browser` and `node` outputs
  - Outputs `esm`, `umd` and `system` + minified versions for browsers
  - Outputs `cjs` for node
- Multiple instance support due to being able to supply a prefix to storage (multiple different Directus installations with their own auth)
- Adds system collection types 
- Adds support for user collection types
- Adds graphql endpoint
- Adds extensions endpoints (typings tbd)
- Tests 
  - Switch to `jest`
  
This adds a new package `sdk` so we can install the sdk using `@directus/sdk` instead. 

---

Closes #3509, closes #3445, closes #3193, closes #4392, closes #3476, closes #4535, closes #4678
Closes #3444 for system tables, and user collections only if given proper typings (client side only)
Might also close #3797 #4386  (needs further understanding)

WoLfulus committed:
add new sdk

WoLfulus commented:
I'd love some reviews/feedback since we have several linked issues/PRs and wanted to let you all know since some ideas came or is really similar to other open PRs too.

@eliashussary @apstanisic

apstanisic commented:
I'll close #3193, this PR solves it completely. I think that this also closes #3476

WoLfulus commented:
Yeah it does! I was aware of that one, but for some reason it didn't land in the description

WoLfulus committed:
update version

eliashussary commented:
> I'd love some reviews/feedback since we have several linked issues/PRs and wanted to let you all know since some ideas came or is really similar to other open PRs too.
> 
> @eliashussary @apstanisic
> 
> ---
> 
> Note that the development of this new SDK is inside the SDK package itself (npm install && npm run dev). I don't know yet how it will play with lerna and we are discussing on changing the way it works right now too.

I've been able to glance over it this morning. Love the direction. I can do a deeper dive on it today. Looks great so far!

WoLfulus committed:
fixes and sdk documentation

WoLfulus committed:
typing updates, documentation

WoLfulus committed:
added missing endpoints

nickrum commented:
The changes to the sdk itself look great.

A few things I noticed regarding the build process:

- It looks like the `tsconfig.types.json` file is unused.
- The browser builds are now built for ES2015 which is not supported by IE11 as far as I know. Should IE support be dropped?
- It seems like all versions are minified, not just the ones ending with `.min.js`.
- Previously, the "legacy" browser build used the iife format instead of umd. Is the amd format included in umd still in use? Otherwise iife could be used for a small file size reduction.
- Is the SystemJS format actually in use as a standalone module loader? I was under the impression it is more like a thing client side bundlers compile esm to for legacy browsers.
- I think the esm build targeting browsers should be separated from the one targeting bundlers as resolving dependencies/minification/polyfilling should likely be handled by the client side bundler.
- Should there also be a node esm build to be used directly within node versions that natively support esm?
- Previously, the iife build was the entrypoint for unpkg. Not sure if this should be changed to the esm build as it seems like this is still not that heavily used outside of bundlers. Also, entrypoint should probably be the minified version.
- Previously, a higher target ES version was used for certain entrypoints to reduce file size and also to let client side bundlers decide about using a lower target ES version.

rijkvanzanten commented:
> The browser builds are now built for ES2015 which is not supported by IE11 as far as I know. Should IE support be dropped?

Calling it, IE 11 is _gone_

nickrum commented:
@rijkvanzanten I'd hope Microsoft would be as clear about it...

rijkvanzanten commented:
Microsoft announced that they're stopping to support IE 11 in Teams and other 365 apps _this year_ (2021), so I think it's safe to say that it's time to start saying farewell 🙂  https://techcommunity.microsoft.com/t5/microsoft-365-blog/microsoft-365-apps-say-farewell-to-internet-explorer-11-and/ba-p/1591666

WoLfulus committed:
targeting minified version for unpkg

WoLfulus committed:
removed unused types file

WoLfulus committed:
fixed non minified versions

WoLfulus commented:
@nickrum thank you! I've updated the PR with what you mentioned. 

> Also, entrypoint should probably be the minified version

I've changed this for unpkg, but I kept the non minified version for "module" as it will probably be used by a bundler later.

> Is the SystemJS format actually in use as a standalone module loader? I was under the impression it is more like a thing client side bundlers compile esm to for legacy browsers.

What you mean by in use? I'm not referencing it in package.json if that's what you're saying. Package.json contains only esm outputs.

> Should there also be a node esm build to be used directly within node versions that natively support esm?

I tried, but the main problem I faced is getting the usage right and still supporting all of them on the same build. I haven't found a proper output to cover all the possible scenarios and yet keep it as simple as `import { Directus } from "@directus/sdk";`

> Previously, a higher target ES version was used for certain entrypoints to reduce file size and also to let client side bundlers decide about using a lower target ES version.

In this regard I just carried what I saw in the old sdk. We can change this. 

Send me a message on Discord or ping me on #sdk if you like to chat about it. We can register what we decide here later.

--- 

Just thinking out loud: it would be nice if we could somehow test the single file builds somehow eventually? With cypress or something else. I'm unsure how redundant this is, but god knows what these bundlers end up with :P

WoLfulus committed:
fix sdk exports

WoLfulus committed:
fix the fix

eliashussary commented on packages/sdk/package.json:
```diff
@@ -0,0 +1,70 @@
+		"url": "https://github.com/directus/directus.git"
+	},
+	"main": "./dist/index.js",
+	"module": "./dist/sdk.esm.js",
```
we're defining a module entrypoint, and defining `axios` as a `dependency`, but based on your rollup output we are also pre-bundling `axios`... 

if this is intended, then it's better to move `axios` to a `devDependency` as it's already being included part of your dist output and there is no need for it on another dev's machine

normally you would leave this to the end-consumer and let their bundle handle it.

the only time you _should_ bundle is if you're shipping directly to browsers. this might be the case, but at that point i would make that a separate build.



with the introduction of `ITransport` there may be some value in looking to make `fetch` your default transport. this would ultimately remove this dependency and make your package even leaner.

WoLfulus replied:
Yeah, I mentioned using fetch to Rijk a while ago, but I haven't done anything with it yet, so I'll have to take a look at it first, but it makes sense to do this instead. Though I think `axios` would land on `{peer/optional}Dependencies` in this case right?

eliashussary replied:
you know what, i just checked out the `cjs` output... it looks like the `require` is intact there. so we still need `axios` listed as a dependency, otherwise using this in a node server context will fail.

no issues leaving it. the only thing up for debate really is whether we should be pre-bundling... i don't think it makes to do it.

if I for example have an existing react-app and i use `axios` for my API, and then need to integrate with `directus` i would expect to see two versions of `axios` now. one bundled in my app, and one pre-bundled via directus. 

if we do not pre-bundle, then we eliminate this duplication and offload it to the consumer of the module.

WoLfulus replied:
I think bundle/not bundle stuff really depends on what you'll be importing on your project. 

IMHO these outputs shouldn't carry dependencies, so I`ll need to triple check the output. 

Additional outputs can be added with a `.bundle` suffix (as I think we had before). I haven`t added them because I was unsure how it should look like and/or behave.

rijkvanzanten replied:
> IMHO these outputs shouldn't carry dependencies, so I`ll need to triple check the output.

Agreed. Dependencies should be external _except_ for the single minified "plop it in a `<script>` and you're done"  `.min.js` file

rijkvanzanten committed:
Remove old sdk

rijkvanzanten committed:
Remove old sdk docs

rijkvanzanten committed:
Install types for Jest, add npm test

rijkvanzanten committed:
Rely on npm exclusively

rijkvanzanten committed:
Remove examples folder

rijkvanzanten committed:
Move typescript down

benhaynes committed:
Update sdk.md

rijkvanzanten commented on packages/sdk/src/collections.ts:
```diff
@@ -0,0 +1,3 @@
+// TODO: collection typing definition utilities
```
@WoLfulus

rijkvanzanten commented on packages/sdk/src/items.ts:
```diff
@@ -0,0 +1,89 @@
+
+export type Filter<T> = {
+	[K in keyof T]?: FilterOperator<T, K> | string | boolean | number | string[] | object;
+	//[key: string]: FilterOperator<T> | string | boolean | number | string[] | object;
```
```suggestion
```

rijkvanzanten commented on packages/sdk/src/storage.ts:
```diff
@@ -0,0 +1,19 @@
+/*
+export type StoredValue = string | number | boolean | Date | StoredObject | StoredArray;
+
+export type StoredArray = StoredValue[];
+
+export type StoredObject = {
+	[key: string]: StoredValue | StoredArray;
+};
+*/
+
```
```suggestion
```

rijkvanzanten commented on packages/sdk/src/types.ts:
```diff
@@ -0,0 +1,92 @@
+	comment: string;
+};
+
+export type CollectionType = {
+	// TODO: complete
+};
+
+export type FieldType = {
+	// TODO: complete
+};
+
+export type FileType = {
+	// TODO: complete
+};
+
+export type FolderType = {
+	// TODO: complete
+};
+
+export type PermissionType = {
+	// TODO: complete
+};
+
+export type PresetType = {
+	// TODO: complete
+};
+
+export type RelationType = {
+	// TODO: complete
+};
+
+export type RevisionType = {
+	// TODO: complete
+};
+
+export type RoleType = {
+	// TODO: complete
+};
```
@WoLfulus

rijkvanzanten commented on packages/sdk/src/types.ts:
```diff
@@ -0,0 +1,92 @@
+	storage_asset_transform: 'none' | 'all' | 'presets';
+};
+
+export type UserType = {
+	// TODO: complete
+};
```
@WoLfulus

rijkvanzanten reviewed (changes requested):
Couple more TODOs to resolve @WoLfulus 🐺 😄 

Also, lets make sure we have auto-refresh implemented

rijkvanzanten commented:
Regarding auto-refresh, lets enhance `.login()` and `.refresh()` with a flag to indicate that refreshing should be automatic, at which point they can start a timeout based on the API-returned expiration time. We could do something like `.login({ refresh: 'auto' })` and `.refresh('auto')` or something along those lines

rijkvanzanten commented on packages/sdk/src/handlers/tfa.ts:
```diff
@@ -0,0 +1,17 @@
+		this.transport = transport;
+	}
+
+	async enable(password: string): Promise<void> {
+		await this.transport.post('/users/tfa/enable', { password });
+	}
+
+	async disable(otp: string): Promise<void> {
+		await this.transport.post('/users/tfa/disable', { otp });
+	}
```
@WoLfulus Can we update this to match https://github.com/directus/directus/pull/4678 ?

WoLfulus committed:
added auto refresh and requested changes

added more http test calls
fixed typing issue in customized types

WoLfulus committed:
remove unused endpoint

WoLfulus committed:
updated docs

WoLfulus committed:
added singletons, fixed typing issues, added password handlers

WoLfulus committed:
rename graphql function and fixed system endpoint

rijkvanzanten committed:
Remove unused imports, fix build

rijkvanzanten committed:
Merge branch 'main' into sdk-update

nickrum commented:
@WoLfulus sorry for not replying for 3 weeks!!!

> I've changed this for unpkg, but I kept the non minified version for "module" as it will probably be used by a bundler later.

Jup, that's what I was trying to suggest 👍

> What you mean by in use? I'm not referencing it in package.json if that's what you're saying. Package.json contains only esm outputs.

What I mean is whether we should actually provide a SystemJS build. To my knowledge it is more like a "polyfill" module loader for esm. I could be wrong though.

> I tried, but the main problem I faced is getting the usage right and still supporting all of them on the same build. I haven't found a proper output to cover all the possible scenarios and yet keep it as simple as `import { Directus } from "@directus/sdk";`

Have you tried using _conditional exports_ within the `exports` field in `package.json`? Something like this:

```json
"exports": {
  "import": "./main-module.js",
  "require": "./main-require.js"
}
```

---

> it would be nice if we could somehow test the single file builds somehow eventually? With cypress or something else. I'm unsure how redundant this is, but god knows what these bundlers end up with :P

I think the Knex devs are doing at least something to test the different output formats. Given how the js ecosystem at least slowly converges to esm, I am not sure either if this is actually necessary.
